pinot-server/src/main/java/org/apache/pinot/server/starter/helix/IngestionBasedConsumptionStatusChecker.java

```diff
@@ -37,62 +38,81 @@ public abstract class IngestionBasedConsumptionStatusChecker {
 
   // constructor parameters
   protected final InstanceDataManager _instanceDataManager;
-  protected final Set<String> _consumingSegments;
+  protected final Map<String, Set<String>> _consumingSegments;
+  protected final Function<String, Set<String>> _consumingSegmentsSupplier;
 
-  // helper variable
-  private final Set<String> _caughtUpSegments = new HashSet<>();
+  // helper variable, which is thread safe, as the method might be called from multiple threads when the health check
+  // endpoint is called by many probes.
+  private final Set<String> _caughtUpSegments = ConcurrentHashMap.newKeySet();
```

Comment:
Do we need to make it thread safe? I feel we can simply synchronize on `getNumConsumingSegmentsNotReachedIngestionCriteria()`. There should be only one caller

Reply:
We may also make `_caughtUpSegments` a map, and whenever a table has all segments caught up, we can remove the table from both maps.
Currently you'll need to merge all table segments into one in each check

Reply:
good point. `synchronized` should be good enough, as the method is supposed to be called via the health check endpoint. It can be called concurrently, but should rare.